workflow/scripts/osemosys_global/powerplant/main.py

```diff
@@ -0,0 +1,290 @@
+    df_oar_final, df_iar_final = activity_master_end(df_oar_base, df_oar_upstream, df_oar_int, 
+                                                     df_oar_trn, df_int_trn_oar, df_iar_base, 
+                                                     df_iar_trn, df_int_trn_iar, duplicate_techs)
+    
+    # Set capital and fixed powerplant costs.
+    df_costs = costs_pwr(weo_costs, costs_dict)
+    
+    # Set capital and fixed transmission costs.
+    df_trans_capex, df_trans_fix = get_transmission_costs(trn_line, df_oar_final)
+    
+    # Combine and format costs data to output as csv.
+    df_cap_cost_final, df_fix_cost_final = costs_end(weo_regions, df_costs, df_oar_final, 
+                                             df_trans_capex, df_trans_fix)
+    
+    # Set CapacityToActivityUnit.
+    df_capact_final = capact(df_oar_final)
+    
+    # Adjust activity limits if cross border trade is not allowed following user config.
+    df_crossborder_final = activity_transmission_limit(cross_border_trade, df_oar_final)
+ 
+    # Set operational life for powerplant technologies.
+    df_op_life_trn, df_op_life_out = set_op_life(tech_code_dict, df_iar_final, 
+                                       df_oar_final, op_life_dict)
+    
+    # Set operational life for transmission.
+    df_op_life = set_op_life_transmission(df_op_life_trn, df_op_life_out, op_life_dict)
+    
+    # Set annual capacity investment constraints.
+    df_max_cap_invest, df_min_cap_invest = cap_investment_constraints(df_iar_final, 
+                                                                      no_investment_techs)
+    
+    # Creates sets for TECHNOLOGIES and FUELS.
+    if custom_nodes:
+        tech_set = create_sets('TECHNOLOGY', df_oar_final, output_data_dir, custom_techs)
+    else:
+        tech_set = create_sets('TECHNOLOGY', df_oar_final, output_data_dir, [])
+        
+    fuel_set = create_sets('FUEL', df_oar_final, output_data_dir, [])       
+    
+    # Create sets for YEAR, MODE_OF_OPERATION and REGION
+    years_set, mode_list_set, regions_set = output_sets(mode_list)
+    
+    # Alter output csv's based on user defined capacities following user config.
+    if not tech_capacity is None:
+        (tech_set, 
+         fuel_set, 
+         df_max_cap_invest, 
+         df_min_cap_invest, 
+         df_res_cap, 
+         df_iar_final, 
+         df_oar_final, 
+         df_op_life, 
+         df_capact_final, 
+         df_cap_cost_final
+         ) = set_user_defined_capacity(
+             tech_capacity, 
+             op_life_dict, 
+             tech_set, 
+             df_min_cap_invest, 
+             df_max_cap_invest, 
+             df_res_cap,
+             df_iar_final,
+             df_oar_final,
+             fuel_set,
+             df_op_life,
+             df_capact_final,
+             df_cap_cost_final, 
+             df_iar_custom_val, 
+             df_oar_custom_val
+             )
```

Comment:
I agree except for the fact that (currently) the legacy code is build this way and I don't want to spend too much extra time right now to pull this apart since it is functional code and priorities are required elsewhere.